LwM2M migrate to new CoAP API

mike-scott opened the pull request:
This patch moves from the ZoAP API in subsys/net/lib/zoap to
the CoAP API in subsys/net/lib/coap which handles multiple
fragments for sending / receiving data.

NOTE: This patch moves the LwM2M library over to the CoAP APIs
but there will be a follow-up patch which re-writes the content
formatter reader / writers to use net_pkt APIs for parsing
across multiple net buffers. The current implementation assumes
all of the data will land in 1 buffer.

Samples using the library still need a fairly large NET_BUF_DATA_SIZE
setting. (Example: CONFIG_NET_BUF_DATA_SIZE=384)

Commit:
net: lwm2m: move to using the new multi-packet CoAP API

This patch moves from the ZoAP API in subsys/net/lib/zoap to
the CoAP API in subsys/net/lib/coap which handles multiple
fragments for sending / receiving data.

NOTE: This patch moves the LwM2M library over to the CoAP APIs
but there will be a follow-up patch which re-writes the content
formatter reader / writers to use net_pkt APIs for parsing
across multiple net buffers. The current implementation assumes
all of the data will land in 1 buffer.

Samples using the library still need a fairly large NET_BUF_DATA_SIZE
setting. (Example: CONFIG_NET_BUF_DATA_SIZE=384)

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: make release in lwm2m_release_message() optional

Let's rename lwm2m_release_message() to lwm2m_reset_message()
and add a parameter to let the function know whether or not to
release the lwm2m_message resource back to the pool.

By adding the optional release parameter, we can keep the
lwm2m_message but reset the underlying net_pkt / net_buf resources.
This allows us to regenerate the net_pkt after determining
an error has occured.  In this case, we don't want the previously
added net_pkt contents but we do want to maintain the message id,
token, etc.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: fix CREATED response code

Handle LwM2M CREATE op correctly and remove comments.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net: lwm2m: fix error handling in handle_request()

Let's reset the payload and return correct response error codes.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

mike-scott commented:
NOTE: This currently includes "net: coap: parse_option() exit gracefully when no payload" patch which is up for separate review.   Once that is merged / fixed, I will rebase here.

mike-scott commented:
Added the rest of the patches to clean up the error handling where we drop the current contents of net_pkt and generate an error code response instead.

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2452,7 +2532,7 @@ static int handle_request(struct zoap_packet *request,
 		} else if (observe == 1) {
 			/* use token from this request */
-			token = zoap_header_get_token(in.in_zpkt, &tkl);
+			tkl = coap_header_get_token(in.in_cpkt, token);
```
We've already extract tkl and token at line 2463. Should be redundant

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -802,40 +804,32 @@ int lwm2m_init_message(struct lwm2m_message *msg)
+	}
+
+	r = coap_packet_init(&msg->cpkt, pkt, 1, msg->type,
+			       tokenlen, token, msg->code,
```
align to (

mike-scott commented:
Rebased on master branch and addressed @rbtchc comments
I also dropped the CoAP API fix as @rveerama1 will put up a PR for that fix separately.
Diff: https://hastebin.com/tatenadijo.hs

rveerama1 commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2023,15 +2014,88 @@ static int lwm2m_delete_handler(struct lwm2m_engine_obj *obj,
+	return len - NET_UDPH_LEN;
+}
+
+#define COAP_MARKER 0xFF
```
Shall we keep one define in include/net/coap.h for this purpose? because subsys/net/lib/coap/coap.c has also one internal define.

mike-scott replied:
We could add this to include/net/coap.h.   However, this define and the following 2 functions go away in a future patch once I've converted the backend LwM2M content formatters to use multi fragments instead of a contiguous buffer.

rveerama1 commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2023,15 +2014,88 @@ static int lwm2m_delete_handler(struct lwm2m_engine_obj *obj,
+#define COAP_MARKER 0xFF
+
+/* TODO: Remove in favor of using net_buf pointer to payload */
+u8_t *coap_packet_get_payload_ptr(struct coap_packet *cpkt, u16_t *len,
```
Is "coap_packet_get_payload()" api not helpful for your requirement?

mike-scott replied:
This is a temporary function to preserve functionality for the LwM2M content formatters until I re-work them to use multi fragments.  They go away once that's complete.

mike-scott commented:
@rbtchc did you have any other issues with this PR?

rbtchc reviewed (approved):
LGTM
